### ISSUE TITLE
[FR] Refresh MLFLOW_TRACKING_TOKEN

### ISSUE THREAD

**Issue by maciejskorski** [mlflow/mlflow#8823]
### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

As of now, MLFlow does not support the use case of registering a long-building artefact with a bearer token.

To my best understanding, `MLFLOW_TRACKING_TOKEN` is set at the beginning of the client session. 

My proposal is to implement re-reading the token on demand.

Note: Extending token life is not an option – possible only for some use-cases and anyways limited.

### Motivation

> #### What is the use case for this feature?

As a user I want to upload long-building artefacts with bearer tokens.

> #### Why is this use case valuable to support for MLflow users in general?

A common challenge when prototyping on large ML models in the cloud.

> #### Why is this use case valuable to support for your project(s) or organization?

Enables scalability of MLFlow.

> #### Why is it currently difficult to achieve this use case?

Not supported / achievable by hacks.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by WeichenXu123** [comment]
@maciejskorski 

> MLFlow does not support the use case of registering a long-building artefact with a bearer token.

Could you elaborate what's `a long-building arrtifact` ? Could you provide a case to describe it and describe when we need to reload the token ?

**Comment by maciejskorski** [comment]
> Could you provide a case to describe it and describe when we need to reload the token ?

@WeichenXu123 Thanks, I added a code snippet describing the functionality. 

This is a common request, see https://github.com/Galileo-Galilei/kedro-mlflow/issues/357

**Comment by WeichenXu123** [comment]
Thank you! What's the purpose to refresh the token during a mlflow run ? Because token expires ?

**Comment by maciejskorski** [comment]
> Thank you! What's the purpose to refresh the token during a mlflow run ? Because token expires ?

Yes, and let's assume that the token life is not manipulable (it is only to a certain degree, and not under more strict cloud security setups). I want mlflow to **refresh the token on my demand during a single client session**.

> would you also provide sketch code of `get_my_token` ?

Tokens, in my application, will be obtained [through the cloud API or CLI](https://developers.google.com/spectrum-access-system/guides/authorization-and-authentication)

**Comment by WeichenXu123** [comment]
OK. I understand your purpose. Would you file PR then we can discuss further based on your implementation ?

**Comment by maciejskorski** [comment]
> Would you file PR then we can discuss further based on your implementation ?

@WeichenXu123, I did a PR in #8879. 

Note that I am not a member of the dev team, so I applied the minimalistic solution so as to not touch too much of the codebase. Would be happy to hear your opinion. Thanks!

**Comment by harupy** [comment]
@maciejskorski 

```python
os.environ['MLFLOW_TRACKING_TOKEN'] = new_token
```

Does this work?

**Comment by harupy** [comment]
```python
import mlflow
import os
from unittest import mock
import requests

original = requests.Session.request


def patch(*args, **kwargs):
    print(kwargs["headers"])
    return original(*args, **kwargs)


with mock.patch("requests.Session.request", patch) as mock_request:
    os.environ["MLFLOW_TRACKING_TOKEN"] = "foo"
    mlflow.set_tracking_uri("http://localhost:5000")
    mlflow.search_experiments()
    os.environ["MLFLOW_TRACKING_TOKEN"] = "bar"
    mlflow.search_experiments()
```

prints out:

```
{'User-Agent': 'mlflow-python-client/2.4.2.dev0', 'Authorization': 'Bearer foo'}
{'User-Agent': 'mlflow-python-client/2.4.2.dev0', 'Authorization': 'Bearer bar'}
```

**Comment by maciejskorski** [comment]
> Let me know if I'm missing something.

@harupy Thanks! Let's see if I understand the architecture.

1. The structure holding the store data is the global var `_tracking_store_registry`, defined in `mlflow.tracking._tracking_service.utils`.

2. I apparently misunderstood the code, assuming it caches the data upon initialization and then it would need refreshing. 

3. Instead, the data like password, user, tokens **are read on-the-fly from configs**. Furthermore, **these configs can also be changed on the fly**, at least in the use-case presented here (indeed, it works!).

Would you mind looking into #8879 and see if has any extra utility (the main use-case seems well-solved already, but the secondary goal was to make registration transparent and clean)?
